packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -477,60 +432,239 @@ contract ControllerHelper is UniswapControllerHelper, AaveControllerHelper, IERC
+        } else if (
+            ControllerHelperDataType.CALLBACK_SOURCE(_callSource) ==
+            ControllerHelperDataType.CALLBACK_SOURCE.FLASHLOAN_CLOSE_VAULT_LP_NFT
+        ) {
+            ControllerHelperDataType.FlashloanCloseVaultLpNftParam memory data = abi.decode(
+                _calldata,
+                (ControllerHelperDataType.FlashloanCloseVaultLpNftParam)
+            );
 
             // convert flashloaned WETH to ETH
-            IWETH9(weth).withdraw(_amount);
+            IWETH9(ControllerHelperDiamondStorage.getAddressAtSlot(5)).withdraw(_amount);
+
+            IController(ControllerHelperDiamondStorage.getAddressAtSlot(0)).deposit{value: _amount}(data.vaultId);
+
+            IController(ControllerHelperDiamondStorage.getAddressAtSlot(0)).withdrawUniPositionToken(data.vaultId);
+
+            (uint256 wPowerPerpAmountInLp, ) = ControllerHelperUtil.closeUniLp(
+                ControllerHelperDiamondStorage.getAddressAtSlot(6),
+                ControllerHelperDataType.closeUniLpParams({
+                    tokenId: data.tokenId,
+                    liquidity: data.liquidity,
+                    liquidityPercentage: data.liquidityPercentage,
+                    amount0Min: data.amount0Min,
+                    amount1Min: data.amount1Min
+                }),
+                isWethToken0
+            );
 
-            IController(controller).deposit{value: _amount}(data.vaultId);
+            ControllerHelperUtil.checkPartialLpClose(
+                ControllerHelperDiamondStorage.getAddressAtSlot(0),
+                ControllerHelperDiamondStorage.getAddressAtSlot(6),
+                data.vaultId,
+                data.tokenId,
+                data.liquidityPercentage
+            );
 
-            IController(controller).withdrawUniPositionToken(data.vaultId);
+            // close short position using amounts collected from closing LP, withdraw collateralToWithdraw + deposited collateralToFlashloan
+            _closeShortWithAmountsFromLp(
+                data.vaultId,
+                wPowerPerpAmountInLp,
+                data.wPowerPerpAmountToBurn,
+                data.collateralToWithdraw.add(data.collateralToFlashloan),
+                data.limitPriceEthPerPowerPerp
+            );
+        } else if (
+            ControllerHelperDataType.CALLBACK_SOURCE(_callSource) ==
+            ControllerHelperDataType.CALLBACK_SOURCE.FLASHLOAN_REBALANCE_VAULT_NFT
+        ) {
+            // convert flashloaned WETH to ETH
+            IWETH9(ControllerHelperDiamondStorage.getAddressAtSlot(5)).withdraw(_amount);
+
+            (uint256 vaultId, ControllerHelperDataType.RebalanceVaultNftParams[] memory data) = abi.decode(
+                _calldata,
+                (uint256, ControllerHelperDataType.RebalanceVaultNftParams[])
+            );
+
+            // check ownership
+            require(IShortPowerPerp(ControllerHelperDiamondStorage.getAddressAtSlot(2)).ownerOf(vaultId) == _initiator);
```

Comment:
Can we move this to the parent function to be consistent with the other ownership require statements, any reason it was placed here? [#303](https://github.com/opynfinance/squeeth-monorepo/issues/303)